### DISCUSSION TITLE
STM32F411 wake from deepsleep using PA0 problem

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#12758] [STM32 / Pyboard] [answered]
Using the recently posted:
https://github.com/nspsck/STM32F411CEU6_BlackPill_Micropython/tree/main
I can't get the Blackpill to wake from deepsleep using PA0.

Also, if I can't get a RTC  wakeup to work.

I found this discussion:
https://github.com/micropython/micropython/issues/3834
did some of mods, but it reboots about 10-15 seconds later without me touching anything.

... disabling storage mode:
`pyb.usb_mode('VCP')`

didn't help.

Found these comprehensive docs:
https://github.com/peterhinch/micropython-micropower/blob/master/UPOWER.md#2-the-pyboard-d
https://github.com/peterhinch/micropython-micropower/blob/master/HARDWARE.md
(pull-down on PA0 didn't help)

Appears being connected to the USB via rshell is a non-starter.  

Do I have to use the UART?  Any other suggestions?

**Comment by davefes** [comment]
If I just program a blinking LED and shut-down rshell then pulling PA0 high works as expected :)

**Reply by nspsck** [reply, marked as answer]
Ugh... So here is what I did, I typed exactly what [Damien George](https://github.com/dpgeorge) has posted in the REPL:

```python
>>> import stm
>>> stm.mem32[stm.PWR + stm.PWR_CSR] |= 1 << 8 # enable WKUP pin on PA0
>>> machine.deepsleep()
```
It sleeps for as long as I do not connect `PA0` to the `3.3v` or `5v` or any other Pins that might send a `HIGH` signal, i.e. Pins that are used by the USB (if I were connecting it using the USB-port) or USART bridge (`PA9`, `PA10` by default) or any other I2C, SPI devices that could have sent unwanted `HIGH` signals. 

> did some of mods, but it reboots about 10-15 seconds later without me touching anything.

So you might want to check if the wake-up source which `PA0` is connected to has unwanted signals. Also you can try to call the `machine.reset_cause()` function on boot in the `main()`function to see if it is actually caused by `4`, which is equal to `machine.DEEPSLEEP_RESET` to sort out other reasons that could be causing the reset. 

PS. I've tested this on a pirated/clone board of very poor quality (i.e. dfu never worked) and it worked properly. 

Also maybe something worth mentioning: I am using Lfs2 for the filesystem, not that will change a thing, at least logically... And I set `pyb.usb_mode('VCP')` in boot because of it. (Windows can not read from Lfs2 directly, so I disabled MSC to avoid the annoying prompt which tells me to format the disk.)

**Reply by davefes** [reply]
Thank you for the comments.  I will get back to deepsleep() when we (@rkompass) get ulora working on this board.

This board looks like a very good solution for remote PV-powered sensors.

**Reply by davefes** [reply]
Another attempt at deepsleep on the STM32F411.  I would like to use a DS3231 to wake this device up.  The DS3231 alarm interrupt is "active low".  The STM32F411 wants to see a high level on PA0 to wake it up, so I fed the DS3231 SQW output into a JFET inverter then connected to PA0.

From my tests it looks like putting a low on PA0 starts up part or all of the "factory reset" process.  So, does PA0 need to be fed with a tri-state driver?  Even though PA0 is supposed to be in pull-down mode I see the voltage on it sits at about 1V, when there is nothing attached to it.

Could someone tell me how to drive PA0 to get the required deepsleep wake-up behaviour and preferably not mess-around with any other STM32F411 functionality?

**Reply by nspsck** [reply]
PA0 is sadly the User switch, hence if that was pressed during the boot process, it will go through the factory reset process. Sadly, that is also the only WKUP on that mcu. Maybe you can just use the built-in rtc LSE? In case that one is not going to satisfy your usecase, define `MICROPY_HW_HAS_SWITCH `to `0 `and delete the related `USRSW` settings. Also, you can use another board. (not the blackpill)

These are only solutions limited by my own set of knowledge. There could well be other more convenient way to work around this.

**Reply by davefes** [reply]
I am not pressing the button, but the DS3231 is holding that pin low.  I am able to operate the RTC wake, but I wanted the DS3231 to control wake-up times.  I can use the RTC to wake-up every minute and check DS3231 time, but that wastes current.

I will look into your suggestion of `MICROPY_HW_HAS_SWITCH` thanks for that.

**Reply by davefes** [reply]
Being aware of this discussion:
https://github.com/micropython/micropython/issues/3834
I was "locked-into" using PA0. Using:
```py
stm.mem32[stm.PWR + stm.PWR_CSR] |= 1 << 8 # enable WKUP pin on PA0
```
Seems that PC13 might be able to be used on the Pyboard, so ...
before I try to port the C-code in:
https://mischianti.org/stm32-power-saving-wake-up-from-external-source-7/#google_vignette
to Micropython should I be able to use any pin that is interrupt-capable to wake a STM32F411 from deepsleep()?

**Reply by davefes** [reply]
Not wanting to give up on getting deepsleep to work on the STM32F411 with a DS3231 RTC here are some further tests.  

I still don't know if PA0 set up with `stm.mem32[stm.PWR + stm.PWR_CSR] |= 1 << 8 # enable WKUP pin on PA0
` makes it an IRQ or just level sensitive.  I know that holding PA0 high seems to stop it working, C1 is my attempt at removing that restraint.
Also, is seems when PA0 is setup as above that it presents a fairly low resistance value on its input, say 30K.  Originally, I had R1 at 100K and that would only pull PA0 up to about 1V.  So, to get a reasonable high on PA0, when required, R1 had to be 10K.

Now, the selection value for C1, 100nF works as do much higher values.  However, dropping back to 10nF the STM32F411 goes through the "factory-reset" process.  So, it seems that the high pulse presented to PA0 has to be some minimum duration.  The time constant for 10K and 100nF = 1mS.
![20231221_093912](https://github.com/micropython/micropython/assets/78183339/88ab850e-e713-4915-beaa-77be03b32681)

Could anybody further my understanding?
